### PR TITLE
Iox #707 extend waitset and listener for member use case

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
The problem which this PR solves is that it is impossible to change for member variables of an object or call member methods inside of a callback when the object itself is not global. The solution is a pointer to a user value which the user can provide to the listener whenever one attaches an event and which is provided as a second argument to the callback. This could then be used to store a this pointer for instance.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- #707

### CODE REVIEW

File: iceoryx_examples/callbacks/README.md [hunk 1]
```diff
@@ -86,18 +86,28 @@ our `heartbeat` user trigger to print the heartbeat message to the console via a
 callback (`heartbeatCallback`).
 
 ```cpp
```
Comment (elBoberido) [inline iceoryx_examples/callbacks/README.md:88]: we need geoffrey :)

File: iceoryx_examples/callbacks/README.md [hunk 2]
```diff
@@ -184,6 +194,113 @@ Afterward, we reset both caches to start fresh again.
     }
 ```
 
+### ice_callbacks_listener_as_class_member.cpp
```
Comment (elBoberido) [inline iceoryx_examples/callbacks/README.md:197]: this is actually more powerful than just a listener as class member. You can pass virtually everything as pointer to additional context data to the callback, being an integer, mutex or instance of a class

File: iceoryx_examples/callbacks/README.md [hunk 3]
```diff
@@ -184,6 +194,113 @@ Afterward, we reset both caches to start fresh again.
...
+
+This example is identical to the [ice_callbacks_subscriber.cpp](#ice_callbacks_subscriber.cpp) 
+one except that we left out the cyclic heartbeat trigger. The key difference is that
```
Comment (elBoberido) [inline iceoryx_examples/callbacks/README.md:200]: not sure if there needs to be a comma between `one except`

File: iceoryx_examples/callbacks/README.md [hunk 4]
```diff
@@ -184,6 +194,113 @@ Afterward, we reset both caches to start fresh again.
...
+This example is identical to the [ice_callbacks_subscriber.cpp](#ice_callbacks_subscriber.cpp) 
+one except that we left out the cyclic heartbeat trigger. The key difference is that 
+the listener is now a class member and in every callback we would like to change 
+some member variables. To do this we require an additional pointer to the object 
+since the listener requires c function references which do not allow the usage
+of lambdas with capturing. Here we can use the userType feature which allows us 
+to provide the this pointer as additional argument to the callback.
+
+The main function is now pretty short, we instantiate our object of type `CounterClass`
+and call `waitForControlC` which uses the `shutdownSemaphore` like in the
```
Comment (elBoberido) [inline iceoryx_examples/callbacks/README.md:208]: at first I thought you accidentally added a `C`. How about calling this just `wait` ot `waitForShutdown`

File: iceoryx_examples/callbacks/README.md [hunk 5]
```diff
@@ -184,6 +194,113 @@ Afterward, we reset both caches to start fresh again.
...
+    iox::popo::Subscriber<CounterTopic> m_subscriberRight;
+    iox::cxx::optional<CounterTopic> m_leftCache;
+    iox::cxx::optional<CounterTopic> m_rightCache;
+    iox::popo::Listener m_listener;
+};
+```
+
+And their purposes are the same as in the previous example. In the constructor
+we initialize the two subscribers and attach them to our listener. But now we 
+add an additional parameter in the `iox::popo::createEventCallback` the
```
Comment (elBoberido) [inline iceoryx_examples/callbacks/README.md:235]: the __n__

File: iceoryx_examples/callbacks/ice_callbacks_listener_as_class_member.cpp [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        /// to the callback to gain access to the object whenever the callback is called.
+        /// It is not possible to use a lambda with capturing here since they are not convertable to
+        /// a C function pointer.
+        m_listener
+            .attachEvent(m_subscriberLeft,
+                         iox::popo::SubscriberEvent::DATA_RECEIVED,
+                         iox::popo::createEventCallback(onSampleReceivedCallback, *this))
+            .or_else([](auto) {
+                std::cerr << "unable to attach subscriberLeft" << std::endl;
+                std::terminate();
```
Comment (elBoberido) [inline iceoryx_examples/callbacks/ice_callbacks_listener_as_class_member.cpp:62]: could you please use `std::exit(EXIT_FAILURE)`

File: iceoryx_posh/include/iceoryx_posh/internal/popo/event_callback.inl [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+#ifndef IOX_POSH_POPO_EVENT_CALLBACK_INL
+#define IOX_POSH_POPO_EVENT_CALLBACK_INL
+
+namespace iox
+{
+namespace popo
+{
+template <typename OriginType, typename UserType>
+inline EventCallback<OriginType, UserType> createEventCallback(void (&callback)(OriginType* const))
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/event_callback.inl:25]: What do you think about renaming `UserType` to `ContextType` or `ContextData`?

File: iceoryx_posh/include/iceoryx_posh/internal/popo/listener.inl [hunk 8]
```diff
@@ -22,39 +22,56 @@ namespace popo
...
+
 template <typename T>
-inline void translateAndCallTypelessCallback(void* const origin, void (*underlyingCallback)(void* const))
+struct TranslateAndCallTypelessCallback<T, internal::NoType_t>
 {
-    reinterpret_cast<void (*)(T* const)>(underlyingCallback)(static_cast<T*>(origin));
-}
+    static void call(void* const origin, void* const userType, Listener::GenericCallbackPtr_t underlyingCallback)
+    {
+        IOX_DISCARD_RESULT(userType);
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/listener.inl:40]: can be used this way, but once we have a `IOX_MAYBE_UNUSED` macro, that one would fit better

**Review by elBoberido — COMMENTED** [review]
Please link the PR to the respective issue.

In a later PR the C binding could be extended with a `iox_listener_attach_subscriber_event_with_user_type/_context`

File: iceoryx_examples/callbacks/README.md (hunk 5 above)
Comment (FerdinandSpitzschnueffler) [inline iceoryx_examples/callbacks/README.md:235]: Please add a comma after `iox::popo::createEventCallback`

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 9]
```diff
@@ -75,7 +75,11 @@ struct iox_listener_storage_t_
 {
     // the value of the array size is the result of the following formula:
     // sizeof(WaitSet) / 8
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:77]: Listener

File: iceoryx_examples/callbacks/README.md (hunk 5 above)
Comment (elBoberido) [reply iceoryx_examples/callbacks/README.md:235]: I'm dumb, see https://github.com/eclipse-iceoryx/iceoryx/pull/715#discussion_r612400118

File: iceoryx_examples/callbacks/README.md (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_examples/callbacks/README.md:88]: When there is time left I will try to cross this point from my todo list

File: iceoryx_posh/include/iceoryx_posh/internal/popo/event_callback.inl (hunk 7 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/internal/popo/event_callback.inl:25]: noted and will be done in the following pull request

File: iceoryx_examples/callbacks/README.md [hunk 10]
```diff
@@ -194,18 +194,23 @@ Afterward, we reset both caches to start fresh again.
     }
 ```
 
-### ice_callbacks_listener_as_class_member.cpp
+### Additional user defined argument for callbacks (ice_callbacks_listener_as_class_member.cpp)
+
+Here we demonstrate how you can virtually everything as an additional argument to the callbacks.
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_examples/callbacks/README.md:199]: pass/provide/add everything...

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

File: iceoryx_examples/callbacks/README.md (hunk 10 above)
Comment (elBoberido) [reply iceoryx_examples/callbacks/README.md:199]: use everything ...? The bets are in :)

File: iceoryx_examples/callbacks/ice_callbacks_subscriber.cpp [hunk 11]
```diff
@@ -103,18 +103,28 @@ int main()
     });
 
     // attach everything to the listener, from here on the callbacks are called when the corresponding event is occuring
-    listener.attachEvent(heartbeat, heartbeatCallback).or_else([](auto) {
+    listener.attachEvent(heartbeat, iox::popo::createEventCallback(heartbeatCallback)).or_else([](auto) {
         std::cerr << "unable to attach heartbeat event" << std::endl;
         std::terminate();
```
Comment (elBoberido) [inline iceoryx_examples/callbacks/ice_callbacks_subscriber.cpp:108]: please also use `std::exit(EXIT_FAILURE)` instead of `std::terminate` here

**Review by elBoberido — COMMENTED** [review]
please link the PR to the relevant issue

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]

**Pull request merged**